### PR TITLE
Fix adding a new section from the unit and course pages

### PR DESCRIPTION
**Pull request by bethanyaconnor**
Re-adds the ability to open the section creation dialog from a link on the course and unit pages. The behavior is similar to the behavior originally added in https://github.com/code-dot-org/code-dot-org/pull/31819. I also considered opening this dialog on the unit/course page but couldn't decide what the next step would be (reloading the page on that section, redirecting to that section's teacher dashboard, something else?), so I just restored the previous behavior.

This broke due to #45305, which removed this code believing it to no longer be used. I couldn't simply revert that PR though as we need to take participant type into account.

From the course page: 
https://user-images.githubusercontent.com/46464143/200376091-69a137fa-c55e-4a32-abe9-31776d5580e2.mov

From the unit page when the unit is within a course:
https://user-images.githubusercontent.com/46464143/200376109-3c2cd0a1-2037-4d78-b42a-4c2ebda62c5a.mov

From the unit page when the unit is standalone:
https://user-images.githubusercontent.com/46464143/200376115-37619771-b362-44f4-a5aa-a00ec2eda5c4.mov

## Testing story

unit tests, tested the cases in the videos above

## Follow-up work

This bug fix is prep for other changes to the section dropdown on course/unit pages. Look out for another PR in a few days! :)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/templates/teacherDashboard/AddSectionDialog.jsx [hunk 1]
```diff
@@ -12,86 +12,109 @@ import {
...
-    // Provided by Redux
-    isOpen: PropTypes.bool.isRequired,
-    section: sectionShape,
-    beginImportRosterFlow: PropTypes.func.isRequired,
-    setRosterProvider: PropTypes.func.isRequired,
-    setLoginType: PropTypes.func.isRequired,
-    setParticipantType: PropTypes.func.isRequired,
-    handleCancel: PropTypes.func.isRequired,
-    availableParticipantTypes: PropTypes.arrayOf(PropTypes.string).isRequired
+const AddSectionDialog = ({
```
Comment (bethanyaconnor) [inline apps/src/templates/teacherDashboard/AddSectionDialog.jsx:25]: I turned this component into a functional component while I was here. It didn't store any state so it was an easy change.

File: apps/src/templates/teacherDashboard/AddSectionDialog.jsx [hunk 2]
```diff
@@ -12,86 +12,109 @@ import {
...
+  isOpen,
+  section,
+  beginImportRosterFlow,
+  setRosterProvider,
+  setLoginType,
+  setParticipantType,
+  handleCancel,
+  availableParticipantTypes,
+  assignedCourseOffering,
+  asyncLoadComplete
```
Comment (bethanyaconnor) [inline apps/src/templates/teacherDashboard/AddSectionDialog.jsx:35]: These two are the new props

File: dashboard/app/views/home/index.html.haml [hunk 3]
```diff
@@ -15,7 +15,7 @@
 - if current_user && current_user.teacher? && !current_user.accepted_latest_terms?
   - current_user.update_user_tos_version_accept()
   = render partial: 'layouts/implicit_new_user_terms_interstitial'
-- elsif SchoolInfoInterstitialHelper.show_confirmation_dialog?(current_user) || @force_school_info_confirmation_dialog
+- elsif (SchoolInfoInterstitialHelper.show_confirmation_dialog?(current_user) || @force_school_info_confirmation_dialog) && !params[:courseVersionId]
```
Comment (bethanyaconnor) [inline dashboard/app/views/home/index.html.haml:18]: Slightly hacky way of preventing both dialogs appearing at once. Let me know if you have a better idea on how to do this!

File: apps/src/templates/teacherDashboard/AddSectionDialog.jsx (hunk 1 above)
Comment (hannahbergam) [reply apps/src/templates/teacherDashboard/AddSectionDialog.jsx:25]: I realize i have no idea when we should use const vs function() when translating to a functional component! Do we have a stated preference somewhere? Also THANKYOU for doing this functional translation at the same time!!

File: apps/src/sites/studio/pages/home/_homepage.js [hunk 4]
```diff
@@ -42,6 +44,29 @@ function showHomepage() {
     store.dispatch(setMapboxAccessToken(homepageData.mapboxAccessToken));
   }
 
+  // remove courseOfferingId, courseVersionId, and unitId params so that if we
+  // navigate back we don't get the create section dialog again
```
Comment (hannahbergam) [inline apps/src/sites/studio/pages/home/_homepage.js:48]: Good thinking, and thank you for adding a comment to explain

File: apps/src/templates/teacherDashboard/AddSectionDialog.jsx [hunk 5]
```diff
@@ -12,86 +12,109 @@ import {
...
+
+  const {loginType, participantType} = section || {};
+  const title = i18n.newSectionUpdated();
+
+  const getDialogContent = () => {
+    if (!asyncLoadComplete) {
+      return <Spinner size="large" style={{padding: 50}} />;
+    }
+    /*
+    The Participant Type Picker will be skips if someone only have permissions to create sections for one
```
Comment (hannahbergam) [inline apps/src/templates/teacherDashboard/AddSectionDialog.jsx]: nit: typo in 'skips'?

File: apps/src/templates/teacherDashboard/AddSectionDialog.jsx [hunk 6]
```diff
@@ -12,86 +12,109 @@ import {
...
+    type of participants. See teacherSectionsRedux for more details on how this is set up.
+    */
+    if (!participantType) {
```
Comment (hannahbergam) [inline apps/src/templates/teacherDashboard/AddSectionDialog.jsx:58]: I wish there were a more easily understandable phrase than participantType... line 41 is really hard to read and intuitively know what is going on

**Review by hannahbergam — APPROVED** [review]
Left a few small comments but nothing blocking. Thank you for doing this, for adding the videos to show what's going on, and for adding these tests!

File: apps/src/templates/teacherDashboard/AddSectionDialog.jsx (hunk 1 above)
Comment (bethanyaconnor) [reply apps/src/templates/teacherDashboard/AddSectionDialog.jsx:25]: not that I know of and I'm definitely guilty of being inconsistent here 🙈

File: apps/src/sites/studio/pages/home/_homepage.js (hunk 4 above)
Comment (bethanyaconnor) [reply apps/src/sites/studio/pages/home/_homepage.js:48]: I have Erin Bond to thank for this :)

File: apps/src/templates/teacherDashboard/AddSectionDialog.jsx (hunk 6 above)
Comment (bethanyaconnor) [reply apps/src/templates/teacherDashboard/AddSectionDialog.jsx:58]: `participantType` is a key in `section`. The series of negated if statements is kind of confusing to me too but I'm not sure how to make it better. We do want it to essentially be:
- if we haven't loaded data, show the spinner
- if we have loaded data but don't know the participant type, show the `ParticipantTypePicker`
- if have loaded data and do know the participant type, but don't know the login type, show the `LoginTypePicker`
- else show the `EditSectionForm`

I'm definitely open to rearranging this if we can think of a way

**Pull request merged**